src/wam2layers/preprocessing/cmip.py

```diff
@@ -0,0 +1,141 @@
+from functools import cache
+from pathlib import Path
+
+import cftime
+import numpy as np
+import pandas as pd
+import xarray as xr
+
+from wam2layers.config import Config
+from wam2layers.utils.calendar import template_to_files
+
+WAM2CMIP = {  # wam2layers name -> cmip name
+    "q": "hus",
+    "e": "hfls",
+    "tp": "pr",
+    "ps": "ps",
+    "u": "ua",
+    "v": "va",
+    "qs": "huss",
+    "us": "uas",
+    "vs": "vas",
+}
+CMIP2WAM = {var: key for key, var in WAM2CMIP.items()}  # cmip name --> wam2layers name
+
+COORD_NAMES = {
+    "lat": "latitude",
+    "lon": "longitude",
+    "plev": "level",
+}
+
+THREE_HR_VARS = ("pr", "hfls")
+
+SECONDS_6HR = 6 * 3600
+LATENT_HEAT = 2.45e6  # J/kg (at ~20 degC, ranges from 2.5 (0 degC) to 2.4 (50 degC)
+# 1.91846e6*((temp/(temp-33.91))**2)
+
+
+def get_input_data(datetime, config):
+    assert config.level_type == "pressure_levels"
+
+    data = xr.Dataset()
+    q = load_cmip_var("q", datetime, config)  # in kg kg-1
+    u = load_cmip_var("u", datetime, config)  # in m/s
+    v = load_cmip_var("v", datetime, config)  # in m/s
+    levels = common_levels(q["level"], u["level"], v["level"])
+
+    data["q"] = load_cmip_var("q", datetime, config).sel(level=levels)
+    data["u"] = load_cmip_var("u", datetime, config).sel(level=levels)
+    data["v"] = load_cmip_var("v", datetime, config).sel(level=levels)
+    data["ps"] = load_cmip_var("ps", datetime, config)  # in Pa
+
+    # Pressure level input vars
+    data["qs"] = load_cmip_var("qs", datetime, config)  # kg kg-1
+    data["us"] = load_cmip_var("us", datetime, config)  # in m/s
+    data["vs"] = load_cmip_var("vs", datetime, config)  # in m/s
+
+    data["precip"] = load_cmip_var("tp", datetime, config)
+    data["evap"] = load_cmip_var("e", datetime, config)
+
+    data["precip"] = np.maximum(data["precip"], 0) + np.maximum(data["evap"], 0)
+    data["evap"] = np.abs(np.minimum(data["evap"], 0))
+
+    return data
+
+
+@cache
```

Comment:
@Peter9192 as especially in CMIP data we access the same files over and over, adding this `@cache` decorator speeds up the loading significantly.

I also added it to the ERA5 loader to reduce the repeated calls to `xr.open_dataarray`, but the speed up is less significant there.

Reply:
Nice, I used to have something similar. Just make sure to add an inline comment to explain what this does, as scientists using who might be reading/editing this code might not be familiar with it.